app/models/project_relationship.rb

```diff
@@ -0,0 +1,41 @@
+# Join model representing some relationship between two `Project`s.
+class ProjectRelationship < ApplicationRecord
+  with_options class_name: 'Project' do
+    belongs_to :left_project
+    belongs_to :right_project
+  end
+
+  has_many :project_edges, inverse_of: :project_relationship # rubocop:disable Rails/HasManyOrHasOneDependent
+  has_many :projects, through: :project_edges, source: :related_project
+
+  validate :no_inverse_pairs
+  validate :no_self_referential_relationships
+
+  private
+
+  def no_inverse_pairs
+    return unless left_project_id && right_project_id
+
+    return unless
+      self.class.where(
+        left_project_id:  left_project_id,
+        right_project_id: right_project_id
+      ).
+      or(
+        self.class.where(
+          left_project_id:  right_project_id,
+          right_project_id: left_project_id
+        )
+      ).
+      exists?
+
+    errors.add(:base, :taken)
```

Comment:
I couldn't find the :taken error message in the en.yml file

Reply:
This one comes for free from Rails:
https://guides.rubyonrails.org/v6.0.4/i18n.html#error-message-interpolation

Since we're riffing on a uniqueness validation it made sense to use the standard uniqueness validation message.

Reply:
Thanks Miles, I wondered if this might be the case because the tests were passing, but couldn't find the documentation to prove it.